python/pyspark/sql/connect/dataframe.py

```diff
@@ -218,8 +218,13 @@ def head(self, n: int) -> Optional["pandas.DataFrame"]:
         self.limit(n)
         return self.toPandas()
 
-    # TODO(martin.grund) fix mypu
-    def join(self, other: "DataFrame", on: Any, how: Optional[str] = None) -> "DataFrame":
+    # TODO: extend `on` to also be type List[ColumnRef].
```

Comment:
What does this mean?

Reply:
Join condition could be something like `[df.name=df2.name, df.age=df2.age]` which will be mapped to List[ColumnRef].

Current expression system in Connect is not perfect that we cannot convert the expression above into a simple expression. There are mixed ColumnRef, String, Expression, List, etc. and we need to figure out a way to unify all of those into ultimately one single expression. If that is too hard then at least we need to change proto to allow accepting a list of Expression.

I will follow up on the expression support in the Connect.